File: Assets/MLRepo/ml.72201f1e-ba18-487e-adc4-4381ae10e21b.davidmatos_watervolume/WaterVolume/Scripts/VolumeWater.cs
using UnityEngine;
using System.Collections.Generic;
using ML.SDK;

    public class VolumeWater : MonoBehaviour
    {
        //|||||||||||||||||||||||||||||||||||||||||||||| PUBLIC VARIABLES ||||||||||||||||||||||||||||||||||||||||||||||
        [Header("Buoyancy Settings")]
        [SerializeField] private bool doBuoyancy = true;
        [SerializeField] private float buoyancyForce = 8f;
        [SerializeField][Range(0, 1)] private float depthPower = 0f;

        [Header("Splash Settings")]
        [SerializeField] private bool doSplashes = true;
        [SerializeField] private GameObject waterSplashEnterPrefab;
        [SerializeField] private GameObject waterSplashExitPrefab;
        [SerializeField] private GameObject waterSurfaceHeight;

        [Header("Player Buoyancy Settings")]
        [SerializeField] private float playerBuoyancyForce = 4f;   // weaker than normal objects
        [SerializeField][Range(0, 1)] private float playerDepthPower = 0.3f;

        //|||||||||||||||||||||||||||||||||||||||||||||| PRIVATE FUNCTIONS ||||||||||||||||||||||||||||||||||||||||||||||
        private void SpawnWaterSplashPrefab(Vector3 prefabSpawnPosition, bool isEntering)
        {
            GameObject prefabToSpawn = isEntering ? waterSplashEnterPrefab : waterSplashExitPrefab;
            if (prefabToSpawn != null)
            {
                Instantiate(prefabToSpawn, prefabSpawnPosition, Quaternion.identity);
            }
        }

        private void ComputeBuoyancyOnRigidbody(Collider collider, Rigidbody rigidbody, float forceOverride = -1f, float depthOverride = -1f)
        {
            if (waterSurfaceHeight == null || rigidbody == null) return;

            float objectYPosition = collider.bounds.center.y;
            float buoyantForceMass = ((forceOverride > 0 ? forceOverride : buoyancyForce) * rigidbody.mass);
            float underWaterBuoyantForce = Mathf.Clamp01(
                (waterSurfaceHeight.transform.position.y - objectYPosition) *
                (depthOverride >= 0 ? depthOverride : depthPower));
            float buoyancy = buoyantForceMass + (buoyantForceMass * underWaterBuoyantForce);

            rigidbody.AddForce(Vector3.up * buoyancy, ForceMode.Force);
        }

        //|||||||||||||||||||||||||||||||||||||||||||||| UNITY FUNCTIONS ||||||||||||||||||||||||||||||||||||||||||||||
        private void Start()
        {
            // Initialization if needed
        }

        private void Update()
        {
            // Not used (kept for consistency)
        }

        private void OnTriggerEnter(Collider other)
        {
            if (!doSplashes || waterSurfaceHeight == null) return;

            Vector3 prefabSpawnPosition = new Vector3(
                other.transform.position.x,
                waterSurfaceHeight.transform.position.y,
                other.transform.position.z);

            // Check if it's a player
            MLPlayer player = other.gameObject.GetPlayer();
            if (player != null)
            {
                SpawnWaterSplashPrefab(prefabSpawnPosition, true);
                return;
            }

            // Check for rigidbodies or grabbable objects
            bool hasRigidbody = other.GetComponent<Rigidbody>() != null ||
                               other.GetComponentInChildren<Rigidbody>() != null;
            bool hasGrabbable = other.GetComponent(typeof(MLGrab)) as MLGrab != null ||
                               other.GetComponentInChildren(typeof(MLGrab)) as MLGrab != null;

            if (hasRigidbody || hasGrabbable)
            {
                SpawnWaterSplashPrefab(prefabSpawnPosition, true);
            }
        }

        private void OnTriggerExit(Collider other)
        {
            if (!doSplashes || waterSurfaceHeight == null) return;

            Vector3 prefabSpawnPosition = new Vector3(
                other.transform.position.x,
                waterSurfaceHeight.transform.position.y,
                other.transform.position.z);

            // Check if it's a player
            MLPlayer player = other.gameObject.GetPlayer();
            if (player != null)
            {
                SpawnWaterSplashPrefab(prefabSpawnPosition, false);
                return;
            }

            // Check for rigidbodies or grabbable objects
            bool hasRigidbody = other.GetComponent<Rigidbody>() != null ||
                               other.GetComponentInChildren<Rigidbody>() != null;
            bool hasGrabbable = other.GetComponent(typeof(MLGrab)) as MLGrab != null ||
                               other.GetComponentInChildren(typeof(MLGrab)) as MLGrab != null;

            if (hasRigidbody || hasGrabbable)
            {
                SpawnWaterSplashPrefab(prefabSpawnPosition, false);
            }
        }

        private void OnTriggerStay(Collider other)
        {
            if (!doBuoyancy) return;

            List<Rigidbody> rigidbodies = new List<Rigidbody>();

            /*
            MLPlayer player = other.gameObject.GetPlayer();
            if (player != null && player.PlayerRoot != null)
            {
                Rigidbody playerRb = player.PlayerRoot.GetComponent<Rigidbody>();
                if (playerRb != null)
                {
                    ComputeBuoyancyOnRigidbody(other, playerRb, playerBuoyancyForce, playerDepthPower);
                }
            }
            */
            
            // Normal rigidbody checks
            Rigidbody parentRb = other.GetComponent<Rigidbody>();
            if (parentRb != null && !rigidbodies.Contains(parentRb))
                rigidbodies.Add(parentRb);

            Component[] childComponents = other.GetComponentsInChildren(typeof(Rigidbody));
            foreach (Component comp in childComponents)
            {
                Rigidbody rb = comp as Rigidbody;
                if (rb != null && !rigidbodies.Contains(rb))
                {
                    rigidbodies.Add(rb);
                }
            }

            // Apply buoyancy to all found rigidbodies
            foreach (Rigidbody rb in rigidbodies)
            {
                ComputeBuoyancyOnRigidbody(other, rb);
            }
        }
    }

File: Assets/MLRepo/ml.72201f1e-ba18-487e-adc4-4381ae10e21b.davidmatos_watervolume/WaterVolume/Scripts/VolumeWater.lua
Namespace("davidmatos_watervolume");
do -- script VolumeWater 
	
	-- get reference to the script
	local VolumeWater = LUA.script;
	
	--|||||||||||||||||||||||||||||||||||||||||||||| PUBLIC VARIABLES ||||||||||||||||||||||||||||||||||||||||||||||
	--|||||||||||||||||||||||||||||||||||||||||||||| PUBLIC VARIABLES ||||||||||||||||||||||||||||||||||||||||||||||
	--|||||||||||||||||||||||||||||||||||||||||||||| PUBLIC VARIABLES ||||||||||||||||||||||||||||||||||||||||||||||

	--------------- Buoyancy ---------------
	local bool_doBuoyancy = SerializedField("Do Buoyancy", Bool);

	--increase value to make object more buoyant, default 8.
	local number_buoyancyForce = SerializedField("Buoyancy Force", Number);

	--value 0 mean no additional Buoyant Force underwater, 1 mean Double buoyant Force underwater (underwater pressure)
	local number_depthPower = SerializedField("Depth Power", Number);

	--------------- Splashes ---------------
	local bool_doSplashes = SerializedField("Do Splashes", Bool);
	local gameObject_waterSplashEnterPrefab = SerializedField("Water Splash Enter Prefab", GameObject);
	local gameObject_waterSplashExitPrefab = SerializedField("Water Splash Exit Prefab", GameObject);
	local gameObject_waterSurfaceHeight = SerializedField("Surface Height Anchor", GameObject);

	--|||||||||||||||||||||||||||||||||||||||||||||| MAIN LOGIC ||||||||||||||||||||||||||||||||||||||||||||||
	--|||||||||||||||||||||||||||||||||||||||||||||| MAIN LOGIC ||||||||||||||||||||||||||||||||||||||||||||||
	--|||||||||||||||||||||||||||||||||||||||||||||| MAIN LOGIC ||||||||||||||||||||||||||||||||||||||||||||||

	local function SpawnWaterSplashPrefab(prefabSpawnPosition, index)
		if(index == true) then
			Object.Instantiate(gameObject_waterSplashEnterPrefab, prefabSpawnPosition, Quaternion(0, 0, 0, 0));
		else
			Object.Instantiate(gameObject_waterSplashExitPrefab, prefabSpawnPosition, Quaternion(0, 0, 0, 0));
		end
	end

	local function ComputeBouyancyOnRigidbody(collider, rigidbody)
		local number_objectYPosition = collider.bounds.center.y;

		local number_buoyantForceMass = number_buoyancyForce * rigidbody.mass;
		local number_underWaterBuoyantForce = Mathf.Clamp01((gameObject_waterSurfaceHeight.transform.position.y - number_objectYPosition) * number_depthPower);
		local number_buoyancy = number_buoyantForceMass + (number_buoyantForceMass * number_underWaterBuoyantForce);

		rigidbody.AddForce(Vector3(0, number_buoyancy, 0), ForceMode.Force);
	end

	--|||||||||||||||||||||||||||||||||||||||||||||| UNITY FUNCTIONS ||||||||||||||||||||||||||||||||||||||||||||||
	--|||||||||||||||||||||||||||||||||||||||||||||| UNITY FUNCTIONS ||||||||||||||||||||||||||||||||||||||||||||||
	--|||||||||||||||||||||||||||||||||||||||||||||| UNITY FUNCTIONS ||||||||||||||||||||||||||||||||||||||||||||||

	-- start only called at beginning
	function VolumeWater.Start()
		LuaEvents.AddLocal(VolumeWater, "A", SpawnWaterSplashPrefab);
	end

	-- update called every frame
	function VolumeWater.Update()

	end

	--NOTE: The local player is the only one that will trip these
	function VolumeWater.OnTriggerEnter(other)
		if(bool_doSplashes == false) then
			return
		end

		local prefabSpawnPosition = Vector3(other.gameObject.transform.position.x, gameObject_waterSurfaceHeight.transform.position.y, other.gameObject.transform.position.z);

		--if the object that entered the trigger volume is infact a player
		if(other.gameObject.IsPlayer() == true) then
			LuaEvents.InvokeLocalForAll(VolumeWater, "A", prefabSpawnPosition, true); --Invoke SpawnWaterSplashPrefab
		else
			local mlgrabParentTest = other.gameObject.GetComponent(MLGrab);
			local rigidbodyParentTest = other.gameObject.GetComponent(Rigidbody);

			local rigidbodyChildren = other.gameObject.GetComponentsInChildren(Rigidbody);
			local rigidbodyChildrenTest = false;

			if(rigidbodyChildren ~= nil) then
				for index1, rigidbodyChild in ipairs(rigidbodyChildren) do
					if(rigidbodyChild ~= nil) then
						rigidbodyChildrenTest = true;
					end
				end
			end

			local mlgrabChildren = other.gameObject.GetComponentsInChildren(MLGrab);
			local mlgrabChildrenTest = false;

			if(mlgrabChildren ~= nil) then
				for index2, mlgrabChild in ipairs(mlgrabChildren) do
					if(mlgrabChild ~= nil) then
						mlgrabChildrenTest = true;
					end
				end
			end

			--if there is a rigidbody, then do a splash
			if(rigidbodyParentTest ~= nil) or (rigidbodyChildrenTest == true) or (mlgrabParentTest ~= nil) or (mlgrabChildrenTest == true) then
				LuaEvents.InvokeLocalForAll(VolumeWater, "A", prefabSpawnPosition, true); --Invoke SpawnWaterSplashPrefab
			end
		end
	end

	--NOTE: The local player is the only one that will trip these
	function VolumeWater.OnTriggerExit(other)
		if(bool_doSplashes == false) then
			return
		end

		local prefabSpawnPosition = Vector3(other.gameObject.transform.position.x, gameObject_waterSurfaceHeight.transform.position.y, other.gameObject.transform.position.z);

		--if the object that entered the trigger volume is infact a player
		if(other.gameObject.IsPlayer() == true) then
			LuaEvents.InvokeLocalForAll(VolumeWater, "A", prefabSpawnPosition, false); --Invoke SpawnWaterSplashPrefab
		else
			local mlgrabParentTest = other.gameObject.GetComponent(MLGrab);
			local rigidbodyParentTest = other.gameObject.GetComponent(Rigidbody);

			local rigidbodyChildren = other.gameObject.GetComponentsInChildren(Rigidbody);
			local rigidbodyChildrenTest = false;

			if(rigidbodyChildren ~= nil) then
				for index1, rigidbodyChild in ipairs(rigidbodyChildren) do
					if(rigidbodyChild ~= nil) then
						rigidbodyChildrenTest = true;
					end
				end
			end

			local mlgrabChildren = other.gameObject.GetComponentsInChildren(MLGrab);
			local mlgrabChildrenTest = false;

			if(mlgrabChildren ~= nil) then
				for index2, mlgrabChild in ipairs(mlgrabChildren) do
					if(mlgrabChild ~= nil) then
						mlgrabChildrenTest = true;
					end
				end
			end

			--if there is a rigidbody, then do a splash
			if(rigidbodyParentTest ~= nil) or (rigidbodyChildrenTest == true) or (mlgrabParentTest ~= nil) or (mlgrabChildrenTest == true) then
				LuaEvents.InvokeLocalForAll(VolumeWater, "A", prefabSpawnPosition, false); --Invoke SpawnWaterSplashPrefab
			end
		end
	end

	--NOTE: The local player is the only one that will trip these
	function VolumeWater.OnTriggerStay(other)
		if(bool_doBuoyancy == false) then
			return
		end

		local rigidbodyParentTest = other.gameObject.GetComponent(Rigidbody);
		local rigidbodyChildren = other.gameObject.GetComponentsInChildren(Rigidbody);

		local totalRigidbodyList = {};

		if(rigidbodyParentTest ~= nil) then
			table.insert(totalRigidbodyList, rigidbodyParentTest);
		end

		if(rigidbodyChildren ~= nil) then
			for index1, rigidbodyChild in ipairs(rigidbodyChildren) do
				table.insert(totalRigidbodyList, rigidbodyChild);
			end
		end

		for index2, rigidbody in ipairs(totalRigidbodyList) do
			ComputeBouyancyOnRigidbody(other, rigidbody);
		end
	end
end